src/pmdas/llc/llc.c

```diff
@@ -0,0 +1,299 @@
+        PMDA_PMUNITS(0,0,0,0,0,0) }, },
+    // mbm_local
+    { NULL, 
+      { PMDA_PMID(0,1), PM_TYPE_U32, LLC_INDOM, PM_SEM_INSTANT, 
+        PMDA_PMUNITS(0,0,0,0,0,0) }, },
+    // mbm_total
+    { NULL, 
+      { PMDA_PMID(0,2), PM_TYPE_U32, LLC_INDOM, PM_SEM_INSTANT, 
+        PMDA_PMUNITS(0,0,0,0,0,0) }, },
+};
+
+static int	isDSO = 1;		/* =0 I am a daemon */
+static char	*username;
+static char	mypath[MAXPATHLEN];
+static char	llcdir[MAXPATHLEN];
+
+typedef struct
+{
+    float llc_occupancy;
+    unsigned long llc_mbm_local;
+    unsigned long llc_mbm_total;
+} llc_metrics;
+    
+/* command line option handling - both short and long options */
+static pmLongOptions longopts[] = {
+    PMDA_OPTIONS_HEADER("Options"),
+    PMOPT_DEBUG,
+    PMDAOPT_DOMAIN,
+    PMDAOPT_LOGFILE,
+    PMOPT_HELP,
+    PMDA_OPTIONS_END
+};
+
+static pmdaOptions     opts = {
+    .short_options = "D:d:l:?",
+    .long_options = longopts,
+};
+
+/*
+ * callback provided to pmdaFetch which handles one metric and one instance
+ */
+static int
+llc_fetchCallBack(pmdaMetric *mdesc, unsigned int inst, pmAtomValue *atom)
+{
+    unsigned int cache_item = pmID_item(mdesc->m_desc.pmid);
+    llc_metrics *lspm;
+    
+    int sts = pmdaCacheLookup(*llc_indom, inst, 0, (void*)&lspm);
+    if (sts < 0 || sts == PMDA_CACHE_INACTIVE)
+	return PM_ERR_INST;
+
+    switch (cache_item) {
+    case 0:		/* occupancy */
+	atom->f = lspm->llc_occupancy;
+	break;
+    case 1:		/* mbm local */
+	atom->ull = lspm->llc_mbm_local;
+	break;
+    case 2:		/* mbm total */
+	atom->ull = lspm->llc_mbm_total;
+    }
+	
+    return PMDA_FETCH_STATIC;
+}
+
+/*
+ * This routine is called once for each pmFetch(3) operation, so caches
+ * each metric for each multiple mon_L3_XX instance
+ */
+static int
+llc_fetch(int numpmid, pmID pmidlist[], pmResult **resp, pmdaExt *pmda)
+{
+    DIR *dirp;
+    FILE *fp;
+    static char	fn[MAXPATHLEN + 1024];
+    char            linebuf[1024];
+    struct dirent   *dentry;
+    unsigned long llc_occupancy_total;
+    float llc_occupancy = 0;
+    unsigned long llc_mbm_local = 0;
+    unsigned long llc_mbm_total = 0;
+    static const char *l3_sys_cache_size  = "/sys/devices/system/cpu/cpu0/cache/index3/size";
+
+    /*
+     * gather per LLC related statistics from the file
+     * /sys/fs/llc/mon_data/mon_L3_XX
+     */
+    dirp = opendir(llcdir);
+    if (! dirp)
+	return PM_ERR_INST;
+
+    static int l3_cache_size;
+
+    if (!l3_cache_size)
+	{
+	    if ( (fp = fopen(l3_sys_cache_size, "r")) != NULL)
+		{
+		    if ( fgets(linebuf, sizeof(linebuf), fp) != NULL)
+			{
+			    sscanf(linebuf, "%uK\n", &l3_cache_size);
+			    l3_cache_size *= 1024;
+			}
+		    fclose(fp);
+		}
+	}
+
+
+    /*
+     * walk the LLC directory, gather each LLC
+     */
+    while ( (dentry = readdir(dirp)) )
+	{
+	    if (strncmp(dentry->d_name, "mon_L3_", 7))
+		continue;
+
+	    snprintf(fn, sizeof fn, "%s/%s/llc_occupancy", llcdir, dentry->d_name);
```

Comment:
Best practice is to use pmsprintf over snprintf as it guarantees null termination (pmsprintf created after previous security reviews of PCP).